docs/Dialects/FIRRTL/FIRRTLAnnotations.md

```diff
@@ -1458,6 +1458,66 @@ Example:
 }
 ```
 
+### OutputDirAnnotation
+
+| Property   | Type   | Description                             |
+| ---------- | ------ | --------------------------------------- |
+| class      | string | `circt.OutputDirAnnotation`             |
+| dirname    | string | The output directory                    |
+| target     | string | Reference target                        |
+
+Specify the output directory for a module. The target must be a module, and must
+be local.
+
+Example:
+```json
+{
+  "class": "circt.OutputDirAnnotation",
+  "dirname": "verification",
+  "target": "~Foo|Bar"
+}
+```
+
+### OutputDirPrecedenceAnnotation
+
+| Property   | Type    | Description                             |
+| ---------- | ------- | --------------------------------------- |
+| class      | string  | `circt.OutputDirPrecedenceAnnotation`   |
+| name       | string  | The output directory                    |
+| parent     | string  | The parent output directory             |
+
+Specify the "parent" of an output directory.
+
+When Verilog is output, some modules will have user-specified output
+directories. When a module `M` is only instantiated by modules which have a
+common output directory `D`, we can also output module `M` in that same
+directory `D`.
+
+If the module M is instantiated under a set of directories `DS`, then the module
+`M` is placed in the output directory that is the least-common-ancestor (LCA) of
+the directories `DS`. The LCA is identified according to this output directory
+declaration annotation.
+
+The intuition behind this annotation is that modules output in the declared
+directory should only depend on modules which are output in an ancestor
+directory. Then, the LCA can be thought of as the "most specific" output
+directory that still makes a module available at all its instantiation
+sites.
+
+When an output directory isn't explicitly declared, then its parent directory is
+implicitly the default output directory. To explicitly declare that a
+directory's parent is the default output directory, use an empty string as the
+parent.
```

Comment:
I'm not 100% sure I am interpreting this comment right

> Is there any problem that this allows the description of arbitrary directed, cyclic graphs as opposed to only trees?

Regarding cycles, and multiple parents, there are checks in the AssignOutputDirs pass that ensure the precedence "graph/tree" is acyclic and that every directory has a single parent declaration.

Regarding multiple roots in the precedence graph/tree, that's not possible either, the graph/tree has a single root, the "default output directory".

Are you suggesting I rework the precedence annotations, so that it can only express a tree by construction? I am open to suggestions, but I chose this representation so that there wouldn't have to be a single monolithic annotation declaring all output directories in a single shot.

I will s/graph/tree/ in the PR for clarity.